File: rust/src/bin/aoc_2025_01.rs
// Advent of Code, 2025 day 01

use std::io;

const START_POSITION: i32 = 50;
const DIAL_SIZE: i32 = 100;

fn parse_move(line: &str) -> i32 {
    let num: i32 = line[1..].parse().unwrap();
    match &line[..1] {
        "L" => -num,
        "R" => num,
        _ => panic!("Bad move line {line}"),
    }
}

fn run(moves: &[i32]) -> (i32, i32) {
    let mut position: i32 = START_POSITION;
    let mut zero_finishes: i32 = 0;
    let mut zero_passes: i32 = 0;
    for mv in moves {
        zero_passes += (mv.abs()) / DIAL_SIZE;
        let m: i32 = (mv.abs() % DIAL_SIZE) * mv.signum();
        position += m;
        if position % DIAL_SIZE == 0 {
            zero_finishes += 1;
            position = 0;
        } else if position > DIAL_SIZE {
            zero_passes += 1;
            position -= DIAL_SIZE;
        } else if position < 0 {
            zero_passes += i32::from(position != m);
            position += DIAL_SIZE;
        }
    }
    (zero_finishes, zero_passes + zero_finishes)
}

fn main() {
    let moves: Vec<i32> = io::stdin()
        .lines()
        .map(|s| parse_move(&s.unwrap()))
        .collect();

    let (part_a, part_b) = run(&moves);
    println!("{part_a}");
    println!("{part_b}");
}

File: python/src/aoc_2025_01.py
"""Advent of Code 2025 - Day 1."""

from sys import stdin
from typing import Final

START_POSITION: Final[int] = 50
DIAL_SIZE: Final[int] = 100


def parse_move(s: str) -> int:
    """Parse one input line."""
    num = int(s[1:])
    match s[0]:
        case "L":
            return -num
        case "R":
            return num
        case _:
            raise ValueError(f"Bad input '{s}'")


def run(_moves: list[int]) -> tuple[int, int]:
    """Run the moves counting zero finishes and clicks."""
    position: int = START_POSITION
    zero_finishes: int = 0
    zero_passes: int = 0
    for move in moves:
        zero_passes += (abs(move)) // DIAL_SIZE
        m: int = (abs(move) % DIAL_SIZE) * (-1 if move < 0 else 1)
        position += m
        if position % DIAL_SIZE == 0:
            zero_finishes += 1
            position = 0
        if position > DIAL_SIZE:
            zero_passes += 1
            position -= DIAL_SIZE
        if position < 0:
            zero_passes += 1 if position != m else 0
            position += DIAL_SIZE
    return zero_finishes, zero_finishes + zero_passes


if __name__ == "__main__":
    moves: list[int] = [parse_move(s) for s in stdin.readlines()]
    part_a, part_b = run(moves)
    print(part_a)
    print(part_b)
